plugins/uefi/fu-uefi-devpath.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2018-2019 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <efivar.h>
+
+#include "fu-uefi-devpath.h"
+
+#include "fwupd-error.h"
+
+typedef struct {
+	guint8	 type;
+	guint8	 subtype;
+	GBytes	*data;
+} FuUefiDevPath;
+
+static void
+fu_uefi_efi_dp_free (FuUefiDevPath *dp)
+{
+	if (dp->data != NULL)
+		g_bytes_unref (dp->data);
+	g_free (dp);
+}
+
+GBytes *
+fu_uefi_devpath_find_data (GPtrArray *dps, guint8 type, guint8 subtype, GError **error)
+{
+	for (guint i = 0; i < dps->len; i++) {
+		FuUefiDevPath *dp = g_ptr_array_index (dps, i);
+		if (dp->type == type && dp->subtype == subtype)
+			return dp->data;
+	}
+	g_set_error (error,
+		     FWUPD_ERROR,
+		     FWUPD_ERROR_INTERNAL,
+		     "no DP with type 0x%02x and subtype 0x%02x",
+		     type, subtype);
+	return NULL;
+}
+
+GPtrArray *
+fu_uefi_devpath_parse (const guint8 *buf, gsize sz,
+		       FuUefiDevpathParseFlags flags, GError **error)
+{
+	guint16 offset = 0;
+	g_autoptr(GPtrArray) dps = NULL;
+
+	/* sanity check */
+	if (sz < sizeof(efidp_header)) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_INTERNAL,
+				     "const_efidp is corrupt");
+		return NULL;
+	}
+
+	dps = g_ptr_array_new_with_free_func ((GDestroyNotify) fu_uefi_efi_dp_free);
+	while (1) {
+		FuUefiDevPath *dp;
+		guint16 hdr_length_fixed = 0;
```

Comment:
It seems like `hdr_length_fixed` should be initialized to `GUINT16_FROM_LE(hdr->length);` directly at this time rather than set to 0 and reset it.

Reply:
Agreed, thanks.